From davem@davemloft.net Tue Apr  7 17:24:11 2009
Return-Path: <davem@davemloft.net>
X-Original-To: dennis@ausil.us
Delivered-To: dennis@ausil.us
Received: from localhost (unknown [127.0.0.1])
	by mail.ausil.us (Postfix) with ESMTP id 8067B158003
	for <dennis@ausil.us>; Tue,  7 Apr 2009 22:24:22 +0000 (UTC)
X-Virus-Scanned: amavisd-new at example.com
Received: from mail.ausil.us ([127.0.0.1])
	by localhost (anubis.ausil.us [127.0.0.1]) (amavisd-new, port 10024)
	with ESMTP id Wtdx2a8O54+g for <dennis@ausil.us>;
	Tue,  7 Apr 2009 17:24:21 -0500 (CDT)
X-Greylist: from auto-whitelisted by SQLgrey-1.7.5
Received: from sunset.davemloft.net (unknown [74.93.104.97])
	by mail.ausil.us (Postfix) with ESMTP id 7A828158001
	for <dennis@ausil.us>; Tue,  7 Apr 2009 17:24:21 -0500 (CDT)
Received: from localhost (localhost [127.0.0.1])
	by sunset.davemloft.net (Postfix) with ESMTP id CA9B8C8D935;
	Tue,  7 Apr 2009 15:24:11 -0700 (PDT)
Date: Tue, 07 Apr 2009 15:24:11 -0700 (PDT)
Message-Id: <20090407.152411.198483029.davem@davemloft.net>
To: dennis@ausil.us
Cc: sparclinux@vger.kernel.org
Subject: Re: kernel bug with CONFIG_KEYBOARD_ATKBD=y
From: David Miller <davem@davemloft.net>
In-Reply-To: <20090407.151638.93201206.davem@davemloft.net>
References: <20090407.145755.213806619.davem@davemloft.net>
	<200904071710.09685.dennis@ausil.us>
	<20090407.151638.93201206.davem@davemloft.net>
X-Mailer: Mew version 6.2.51 on Emacs 22.1 / Mule 5.0 (SAKAKI)
Mime-Version: 1.0
Content-Type: Text/Plain;
  charset=us-ascii
Content-Transfer-Encoding: 7bit
X-UID: 4807
X-Length: 3606
Status: R
X-Status: N
X-KMail-EncryptionState:  
X-KMail-SignatureState:  
X-KMail-MDN-Sent:  

From: David Miller <davem@davemloft.net>
Date: Tue, 07 Apr 2009 15:16:38 -0700 (PDT)

> From: Dennis Gilmore <dennis@ausil.us>
> Date: Tue, 7 Apr 2009 17:10:08 -0500
> 
>> On Tuesday 07 April 2009 04:57:55 pm David Miller wrote:
>>> From: Dennis Gilmore <dennis@ausil.us>
>>> Date: Tue, 7 Apr 2009 10:01:17 -0500
>>>
>>> > Initalizing network drop monitor service
>>> > kernel BUG at
>>> > /builddir/build/BUILD/kernel-2.6.29/linux-2.6.29.sparc64/arch/sparc/inclu
>>> >de/asm/tlb_64.h:48!
>>>
>>> What kernel is this tree based upon?  2.6.29.1?
>> yeah 2.6.29.1
> 
> A sparc64 fix I put into 2.6.29.1 is likely the cause.  It's possible
> I put in a version without a particular issue cured.

Indeed, here is the fix, I'll push it to -stable.  Thanks for your
report:

sparc64: Fix bug in ("sparc64: Flush TLB before releasing pages.")

[ No upstream commit, this regression was added only to 2.6.29.1 ]

Unfortunately I merged an earlier version of commit
b6816b706138c3870f03115071872cad824f90b4 ("sparc64: Flush TLB before
releasing pages.") than what I actually tested and merged upstream.

Simply diffing asm/tlb_64.h in Linus's tree vs. what ended up in
2.6.29.1 confirms this.

Sync things up to fix BUG() triggers some users are seeing.

Reported-by: Dennis Gilmore <dennis@ausil.us>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 arch/sparc/include/asm/tlb_64.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/sparc/include/asm/tlb_64.h b/arch/sparc/include/asm/tlb_64.h
index 0aaa086..ee38e73 100644
--- a/arch/sparc/include/asm/tlb_64.h
+++ b/arch/sparc/include/asm/tlb_64.h
@@ -57,9 +57,9 @@ static inline struct mmu_gather *tlb_gather_mmu(struct mm_struct *mm, unsigned i
 
 static inline void tlb_flush_mmu(struct mmu_gather *mp)
 {
+	if (!mp->fullmm)
+		flush_tlb_pending();
 	if (mp->need_flush) {
-		if (!mp->fullmm)
-			flush_tlb_pending();
 		free_pages_and_swap_cache(mp->pages, mp->pages_nr);
 		mp->pages_nr = 0;
 		mp->need_flush = 0;
-- 
1.6.2.2

